cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java

```diff
@@ -266,15 +268,42 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
 
     service.setSpec(spec);
 
-    try {
-      api.createNamespacedService(namespace, service, null, null, null);
-      LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
-    } catch (ApiException e) {
-      // It means the service already exists. In this case we update the port if it is not the same.
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
-        return false;
+    // Maximum elapse-able time currently set to 10 minutes
+    ExponentialBackOff backoffGenerator
+        = new ExponentialBackOff.Builder().setMaxElapsedTimeMillis(600000).build();
+    boolean serviceCreatedSuccessfully = false;
+    long backoffDuration = ExponentialBackOff.DEFAULT_INITIAL_INTERVAL_MILLIS;
+
+    while (!serviceCreatedSuccessfully) {
+      try {
+        api.createNamespacedService(namespace, service, null, null, null);
+        LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
+        serviceCreatedSuccessfully = true;
+      } catch (ApiException e) {
+        // It means the service already exists. In this case we update the port if it is not the same.
+        if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
+          return false;
+        } else {
+          if (e.getCause() instanceof SocketTimeoutException) {
+            LOG.warn("Service creation in kubernetes with name {} and port {} encountered exception: {}",
+                     serviceName,
+                     port.getPort(),
+                     e);
+            try {
+              Thread.sleep(backoffDuration);
+              backoffDuration = backoffGenerator.nextBackOffMillis();
+            } catch (Exception f) {
+              LOG.warn("Encountered exception during k8s service creation backoff: " + f);
```

Comment:
What kind of exception would it be? Is it interrupted exception? If you ignore this, it means the caller may be blocked inside this method and not able to cancel it (e.g. upon an explicit shutdown call).